runtime/ffi/ffi.cpp

```diff
@@ -0,0 +1,200 @@
+    } else if (symbol == tag_type_uint64()) {
+      return ffi_type_uint64;
+    } else if (symbol == tag_type_sint64()) {
+      return ffi_type_sint64;
+    } else if (symbol == tag_type_float()) {
+      return ffi_type_float;
+    } else if (symbol == tag_type_double()) {
+      return ffi_type_double;
+    } else if (symbol == tag_type_uchar()) {
+      return ffi_type_uchar;
+    } else if (symbol == tag_type_schar()) {
+      return ffi_type_schar;
+    } else if (symbol == tag_type_ushort()) {
+      return ffi_type_ushort;
+    } else if (symbol == tag_type_sshort()) {
+      return ffi_type_sshort;
+    } else if (symbol == tag_type_uint()) {
+      return ffi_type_uint;
+    } else if (symbol == tag_type_sint()) {
+      return ffi_type_sint;
+    } else if (symbol == tag_type_ulong()) {
+      return ffi_type_ulong;
+    } else if (symbol == tag_type_slong()) {
+      return ffi_type_slong;
+    } else if (symbol == tag_type_longdouble()) {
+      return ffi_type_longdouble;
+    } else {
+      throw std::invalid_argument("Arg is not a supported type");
+    }
+  }
+
+  block * hook_FFI_call(mpz_t addr, struct list * args, struct list * types, block * ret) {
+    ffi_cif cif;
+    ffi_type *argtypes, rtype;
+    void (* address)();
+
+    if (!mpz_fits_ulong_p(addr)) {
+      throw std::invalid_argument("Addr is too large");
+    }
+    *(void **)(&address) = (void *) mpz_get_ui(addr);
+
+    mpz_ptr nargs_ptr = hook_LIST_size(args);
+    mpz_ptr ntypes_ptr = hook_LIST_size(types);
+    if (!mpz_fits_ulong_p(nargs_ptr)) {
+      throw std::invalid_argument("Args list is too large");
+    }
+    if (!mpz_fits_ulong_p(ntypes_ptr)) {
+      throw std::invalid_argument("Types list is too large");
+    }
+
+    size_t nargs = mpz_get_si(nargs_ptr);
+    size_t ntypes = mpz_get_si(ntypes_ptr);
+
+    if (nargs != ntypes) {
+      throw std::invalid_argument("Args size does not match types size");
+    }
+
+    argtypes = (ffi_type *) malloc(sizeof(ffi_type) * nargs);
+
+    block * type, * elem;
+    ffi_type arg_t;
+    size_t avalues_size = 0;
+    for (int i = 0; i < nargs; i++) {
+        elem = hook_LIST_get(types, i);
+        if (elem->h.hdr != (uint64_t)getTagForSymbolName("inj{SortFFIType{}}")) {
+          throw std::invalid_argument("Types list contains invalid FFI type");
+        }
+        type = (block *) *elem->children;
+        arg_t = getTypeFromSymbol((uint64_t)type);
+        argtypes[i] = arg_t;
+        avalues_size += arg_t.size;
+    }
+
+    string * bytes;
+    void * avalues = malloc(avalues_size);
+    size_t len, offset = 0;
+    for (int i = 0; i < nargs; i++) {
+        elem = hook_LIST_get(args, i);
+        if (elem->h.hdr != (uint64_t)getTagForSymbolName("inj{SortBytes{}}")) {
+          throw std::invalid_argument("Args list contains non-bytes type");
+        }
+        bytes = (string *) *elem->children;
+        len = len(bytes);
+        memcpy((char *)avalues + offset, bytes, sizeof(char) * len);
```

Comment:
you should not need to memcpy here. Since it is expecting a pointer to an array of bytes, and you have a pointer to an array of bytes, you can just write the pointer to the array of pointers.